In [5]:
#importing libraries
import requests
from bs4 import BeautifulSoup
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [6]:
#estabilishing connection with web page address
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.108 Safari/537.36", "Accept-Encoding":"gzip, deflate", "Accept":"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8", "DNT":"1","Connection":"close", "Upgrade-Insecure-Requests":"1"
}

In [17]:
#scrapping the data 
data = {"title":[],"review":[], "year":[]}
for x in range(1,76):
  pages = requests.get(f'https://www.amazon.com/s?k=amazon+shopping+online&i=stripbooks&rh=n%3A283155&dc&page={x}&adgrpid=84486094307&gclid=CjwKCAjwwab7BRBAEiwAapqpTDYtTN8aym2q_rlYn1VdmofKw0_bzkd6xjaCl8cMY4syjbPZMKNMjxoCLIYQAvD_BwE&hvadid=393524136919&hvdev=c&hvlocphy=1010294&hvnetw=g&hvqmt=b&hvrand=2170543926145783317&hvtargid=kwd-469356929&hydadcr=22365_10729094&qid=1635883021&rnid=2941120011&ref=sr_pg_{x}',
                        headers=headers)
  soup = BeautifulSoup(pages.content, "html.parser")

  review=soup.find_all('div', {"class": "a-row a-size-small"})
  for i in review:
    for j in i.find_all('span', {"class": "a-size-base"}):
      data["review"].append(j.text)
      


  title = soup.find_all('h2', {"class": "a-size-mini a-spacing-none a-color-base s-line-clamp-2"})
  for titles in title:
    data["title"].append(titles.text)

  for year in soup.find_all('span', { 'class':'a-size-base a-color-secondary a-text-normal' }):
      data["year"].append(year.text)

In [18]:
#storing the data in a dataframe
books= pd.DataFrame.from_dict(data, orient='index')   
books.T

title  review          year
0     Luxury Bridal Satin Pillow Case with Concealed...  34,683   Dec 5, 2019
1     Shit I Can’t Remember: password book, password...     782   Oct 1, 2019
2     Publication Manual of the American Psychologic...  22,720   Jul 8, 2020
3     Amazon FBA, Dropshipping Shopify, Social Media...     109   Jun 1, 2015
4     Old World Internet Address & Password Logbook ...   2,326  Feb 23, 2019
...                                                 ...     ...           ...
1632  Online Shopping Tracker: Organize your purchas...    None          None
1633  The Pocket Stylist: Behind-the-Scenes Expertis...    None          None
1634  This Is Why I'm Broke An Online Shopping Track...    None          None
1635  Online Shopping Tracker: Keep Track Your Onlin...    None          None
1636  EL METABOLISMO Y EL EFECTO REBOTE: Una guía co...    None          None

[1637 rows x 3 columns]

In [19]:
#saving the stored data into a csv file 
books.T.to_csv('books.csv', index=False) 

In [20]:
#loading the data
df= pd.read_csv("books.csv")
df

title  review          year
0     Luxury Bridal Satin Pillow Case with Concealed...  34,683   Dec 5, 2019
1     Shit I Can’t Remember: password book, password...     782   Oct 1, 2019
2     Publication Manual of the American Psychologic...  22,720   Jul 8, 2020
3     Amazon FBA, Dropshipping Shopify, Social Media...     109   Jun 1, 2015
4     Old World Internet Address & Password Logbook ...   2,326  Feb 23, 2019
...                                                 ...     ...           ...
1632  Online Shopping Tracker: Organize your purchas...     NaN           NaN
1633  The Pocket Stylist: Behind-the-Scenes Expertis...     NaN           NaN
1634  This Is Why I'm Broke An Online Shopping Track...     NaN           NaN
1635  Online Shopping Tracker: Keep Track Your Onlin...     NaN           NaN
1636  EL METABOLISMO Y EL EFECTO REBOTE: Una guía co...     NaN           NaN

[1637 rows x 3 columns]

In [21]:
#inspecting the dataframe
df.head(20)

title  review          year
0   Luxury Bridal Satin Pillow Case with Concealed...  34,683   Dec 5, 2019
1   Shit I Can’t Remember: password book, password...     782   Oct 1, 2019
2   Publication Manual of the American Psychologic...  22,720   Jul 8, 2020
3   Amazon FBA, Dropshipping Shopify, Social Media...     109   Jun 1, 2015
4   Old World Internet Address & Password Logbook ...   2,326  Feb 23, 2019
5   Everything You Need to Know When I'm Gone - En...   1,226   Dec 6, 2018
6   Thrift Store Reselling Secrets You Wish You Kn...     258  Aug 24, 2021
7               Columbia Men's Watertight Ii Jacket    18,391  Oct 26, 2019
8   EL METABOLISMO Y EL EFECTO REBOTE: Una guía co...   4,127  Jan 13, 2014
9   WTF Is My Password: Password Book Log Book Alp...     143  Oct 14, 2021
10  Shopping Addiction: The Ultimate Guide for How...      68  May 10, 2020
11  We Are All Related: Online Shopping and Gift I...      50   Sep 4, 2018
12  Pre-Owned Clothing Business Startup: How to St...      97  Nov 28, 2019
13  Low Carb Dieta: Recetas para principiantes Guí...   5,398  Aug 23, 2021
14  The Secrets to Selling on Amazon: How I Turned...     135   Jan 1, 1975
15  The Game of Internet Marketing (Bundle): Creat...     633  Sep 21, 2021
16            A Course in Miracles: Combined Volume         8   Dec 8, 2020
17      FOODHEIM: A Culinary Adventure [A Cookbook]     2,043   Aug 7, 2021
18     Keto Cocktails: Keto Diet Cookbook Cocktails         1  Dec 15, 2014
19                 CONQUEROR: How I lost 100 Pounds     1,864   Oct 3, 2021

In [22]:
#inspecting the dataframe
df.tail(20)

title review year
1617  Low Carb Dieta: Recetas para principiantes Guí...    NaN  NaN
1618            Your No-Nonsense Guide to Eating Well      NaN  NaN
1619  Stedman's Medical Terminology: Steps to Succes...    NaN  NaN
1620  Online shopping tracker log book: 120 page 6 X...    NaN  NaN
1621  Glow15: A Science-Based Plan to Lose Weight, R...    NaN  NaN
1622  Puglia Travel Guide (Quick Trips Series): Sigh...    NaN  NaN
1623  Making A List, Checking It Twice: Online Shopp...    NaN  NaN
1624                      Dr. Atkins' Diet Revolution      NaN  NaN
1625                       Frommer's Paris day by day      NaN  NaN
1626  Bacon & Butter: The Ultimate Ketogenic Diet Co...    NaN  NaN
1627  Don't Sugar Coat It: The 21 Day Sugar Free Diet      NaN  NaN
1628   Expiry Date (The Ant and Bea Mysteries Book 3)      NaN  NaN
1629  The Healthy Gluten-Free Life: 200 Delicious Gl...    NaN  NaN
1630  Reseller Inventory Log: Product Listing Notebo...    NaN  NaN
1631  Christmas Planner: Day planner, Recipe Section...    NaN  NaN
1632  Online Shopping Tracker: Organize your purchas...    NaN  NaN
1633  The Pocket Stylist: Behind-the-Scenes Expertis...    NaN  NaN
1634  This Is Why I'm Broke An Online Shopping Track...    NaN  NaN
1635  Online Shopping Tracker: Keep Track Your Onlin...    NaN  NaN
1636  EL METABOLISMO Y EL EFECTO REBOTE: Una guía co...    NaN  NaN

In [23]:
#concise summary of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1637 entries, 0 to 1636
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   title   1637 non-null   object
 1   review  1315 non-null   object
 2   year    1582 non-null   object
dtypes: object(3)
memory usage: 38.5+ KB


In [24]:
#checking for missing values
df.isna().sum()

title       0
review    322
year       55
dtype: int64

There are 322 books without reviews and 55 books without publication date.we remove rows with missing values.

In [27]:
#dropping missing data
df.dropna(axis=0, inplace= True)

In [28]:
#confirming if there are missing values left
df.isna().sum()

title     0
review    0
year      0
dtype: int64

In [34]:
# extracting only the year values from the year column
df['year'] = df['year'].str[-4:]

#removing "," from the reviews column
df['review'] = df['review'].str.replace(",","")

In [35]:
#converting year and review columns to integers
df['year'] = df['year'].astype(int)
df['review'] = df['review'].astype(int)

In [41]:
#removing unwanted data(ad that was scrapped with data)
df = df.iloc[1: , :]

In [36]:
#concise summary of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1315 entries, 0 to 1314
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   title   1315 non-null   object
 1   review  1315 non-null   int64 
 2   year    1315 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 41.1+ KB


Extracting books that were up for sale from 2009 to 2019

In [42]:
#creating new dataframe that shows books on sale between  2009 and 2019
df2= df[(df['year'] >= 2009) & (df['year'] <= 2019)]

In [43]:
#checking random samples of the data
df2.sample(20)

title  review  year
1132  Reseller Inventory Log: Product Listing Notebo...    1687  2014
757           Statistics: Unlocking the Power of Data      1725  2010
1039                 CONQUEROR: How I lost 100 Pounds      2043  2018
1312  Low Carb Dieta: Recetas para principiantes Guí...     708  2010
843                    Strength Training for Fat Loss        38  2019
528   Online Shopping Tracker: Online Shopping Track...     126  2017
922   Internet Password Book Organizer And Online Sh...     138  2016
1290  Low Carb Dieta: Recetas para principiantes Guí...       8  2015
712   Just Grandma and Me (Little Critter) (Pictureb...       2  2018
477                           A Simple Favor: A Novel        70  2014
287   Sorry for Your Loss - It’s Me: My Final Though...     160  2011
493                  CONQUEROR: How I lost 100 Pounds      2043  2016
729   Yahoo! Hacks: Tips & Tools for Living on the W...      34  2013
1082  Online Shopping Tracker: Online Order Organize...      38  2014
339   EL METABOLISMO Y EL EFECTO REBOTE: Una guía co...    4975  2018
418   Don't Sugar Coat It: The 21 Day Sugar Free Diet       633  2014
318   Stop Bored Eating: How to Overcome Emotional E...     106  2019
694   Online shopping Tracker Notebook: Keep Track a...       8  2016
398                     Vanished: A Nick Heller Novel      1550  2017
465   Christmas Planner: Festive Organiser : Plan an...       4  2014

assuming the frequency at which a book is bought is relative to the number of reviews the book gets.



In [49]:
#sorting df2 by number of reviews
df2. sort_values("review", ascending= False)

title  review  year
44    Keto Diet for Women Over 50: The 30-Day Journe...   25185  2015
105   Maangchi's Big Book of Korean Cooking: From Ev...   19360  2014
7                 Columbia Men's Watertight Ii Jacket     18391  2019
256   Christmas Planner: Festive Organiser : Plan an...   17325  2017
61    Dotcom Secrets: The Underground Playbook for G...   13445  2019
...                                                 ...     ...   ...
993   Online shopping Tracker: Online shopping Track...       1  2017
867   Password Logbook With Alphabetical Tabs: Organ...       1  2013
1108                                      Qualityland         1  2009
1264                Design of Cities: Revised Edition         1  2018
884   Raw and Simple: Eat Well and Live Radiantly wi...       1  2018

[630 rows x 3 columns]

In [45]:
#statistical summary of the new dataframe
df2.describe(include='all')

title  ...         year
count                                                 630  ...   630.000000
unique                                                426  ...          NaN
top     EL METABOLISMO Y EL EFECTO REBOTE: Una guía co...  ...          NaN
freq                                                   29  ...          NaN
mean                                                  NaN  ...  2015.717460
std                                                   NaN  ...     2.923972
min                                                   NaN  ...  2009.000000
25%                                                   NaN  ...  2014.000000
50%                                                   NaN  ...  2016.000000
75%                                                   NaN  ...  2018.000000
max                                                   NaN  ...  2019.000000

[11 rows x 3 columns]

In [57]:
#sorting the dataframe based on reviews
df2. sort_values("review", ascending= False).head(20)

title  review  year
44    Keto Diet for Women Over 50: The 30-Day Journe...   25185  2015
105   Maangchi's Big Book of Korean Cooking: From Ev...   19360  2014
7                 Columbia Men's Watertight Ii Jacket     18391  2019
256   Christmas Planner: Festive Organiser : Plan an...   17325  2017
61    Dotcom Secrets: The Underground Playbook for G...   13445  2019
57    The Spy and the Traitor: The Greatest Espionag...   13375  2014
313   The Alkaline Reset Cleanse: The 7-Day Reboot f...   13184  2018
136   Make Each Click Count Using Google Shopping: R...   13184  2018
154   Keto Chaffle Cookbook: 500+ Quick and Super Ea...   13184  2013
46                                The Food of Sichuan     12195  2018
1175  Shopping tracker: Online Shopping organizer an...   10262  2019
228   Posh Boss Online Reseller Inventory Tracker Re...    9691  2015
381   The Savage Garden, Revised: Cultivating Carniv...    9691  2012
96    Younger Next Year: Live Strong, Fit, Sexy, and...    8849  2019
720      Keto Cocktails: Keto Diet Cookbook Cocktails      8817  2013
466   Online Shopping Tracker: Online Order Organize...    8803  2018
126                                            Plazas      8740  2010
138   1897 Sears, Roebuck & Co. Catalogue: A Window ...    8740  2013
157   2021 Online Reseller Sourcing Guide: Huge List...    8740  2014
42                   CONQUEROR: How I lost 100 Pounds      8544  2019

**Observation**

Based on the number of reviews, the types of books frequently bought from 2009 to 2019 according to the data scrapped are;

* Diet books: books for weight loss
* cookery books
* Calenders: shopping trackers and christmas planners and
* Novels